### PR TITLE
Dryrun rpc

### PR DESCRIPTION
**Pull request by vivekvpandya**
## Issue
https://app.clickup.com/t/30xdrqf

## Description
RPC method to check if add_liquidity() and remove_liquidity() will success or not.

## Checklist

~- [ ] I have updated the cargo docs to reflect changes made by this PR _(if applicable)_~
~- [ ] I have updated the `book/` to reflect changes made by this PR _(if applicable)_~

### CODE REVIEW

**Review by dzmitry-lahoda — DISMISSED** [review]

File: frame/composable-traits/src/dex.rs
Comment (KaiserKarel) [inline frame/composable-traits/src/dex.rs:40]: If this `/// Returns the amount of base & quote asset redeemable for given amount of lp token.`, why do we need to add min_base and min_quote parameters?

File: frame/composable-traits/src/dex.rs
Comment (KaiserKarel) [inline frame/composable-traits/src/dex.rs:36]: nit: 
```suggestion
	fn redeemable_assets_for_lp_tokens(
```

File: frame/pablo/rpc/src/lib.rs
Comment (KaiserKarel) [inline frame/pablo/rpc/src/lib.rs:37]: Aren't RPCs always dry runs?

File: frame/pablo/rpc/src/lib.rs
Comment (KaiserKarel) [inline frame/pablo/rpc/src/lib.rs:47]: Aren't RPCs always dry runs?

File: frame/pablo/rpc/src/lib.rs
Comment (KaiserKarel) [inline frame/pablo/rpc/src/lib.rs:109]: Aren't RPCs always dry runs?

File: frame/composable-traits/src/dex.rs
Comment (vimukthi-git) [inline frame/composable-traits/src/dex.rs:129]: We need to introduce asset Ids (lists of assets) for these types of calls, names like `base` and `quote` doesn't cut it anymore as we want to move towards multi asset pools.

File: frame/composable-traits/src/dex.rs
Comment (vimukthi-git) [inline frame/composable-traits/src/dex.rs:39]: Same as the other comment

File: frame/dex-router/src/lib.rs
Comment (vimukthi-git) [inline frame/dex-router/src/lib.rs:457]:
```suggestion
		fn simulate_remove_liquidity(
```

File: frame/dex-router/src/lib.rs
Comment (vimukthi-git) [inline frame/dex-router/src/lib.rs:425]:
```suggestion
		fn simulate_add_liquidity(
```

File: frame/composable-traits/src/dex.rs
Comment (vimukthi-git) [inline frame/composable-traits/src/dex.rs:412]:
```suggestion
pub struct RemoveLiquiditySimulationResult<AssetId, Balance>
```

File: frame/pablo/src/lib.rs
Comment (KaiserKarel) [inline frame/pablo/src/lib.rs:726]: I'd prefer for this to be methods on pool, so that the body of this function is reduced:

```
impl PoolConfiguration {
    // Returns number of LP tokens for the given pool
    fn lp_for_liquidity(...) -> Balance
}
```

Then the body can be:

```
let pool = Self::get_pool(pool_id)?;
let pool_account = Self::account_id(&pool_id);
ensure!(T::Assets::reducible_balance(pool.pair.base, who, false) >= base_amount,
Error::<T>::NotEnoughLiquidity
);
ensure!(
T::Assets::reducible_balance(pool.pair.quote, who, false) >= quote_amount,
Error::<T>::NotEnoughLiquidity
);
let lp = pool.lp_for_liquidity(base_amount, quote_amount);
```

File: frame/composable-traits/src/dex.rs
Comment (vimukthi-git) [reply frame/composable-traits/src/dex.rs:129]: Probably best to have a common data structure that captures an `asset_id` and a `balance`

File: frame/composable-traits/src/dex.rs
Comment (vivekvpandya) [reply frame/composable-traits/src/dex.rs:40]: Because we don't want a check for CannotRespectMinimuAmount in frontend. with those two parameters it is done in RPC.

File: frame/composable-traits/src/dex.rs
Comment (vivekvpandya) [reply frame/composable-traits/src/dex.rs:129]: Is if fine to do that in separate change? AFAIK these RPCs are priority for FE.

File: frame/composable-traits/src/dex.rs
Comment (KaiserKarel) [reply frame/composable-traits/src/dex.rs:40]: :+1:

File: frame/composable-traits/src/dex.rs
Comment (vivekvpandya) [reply frame/composable-traits/src/dex.rs]: Also I request to have small PRs which has changes related to the subject of given click-up ticket. This allows reviewers to have fast reviews and make overall progress faster. IMHO diverging to other large subject is not good.

Comment (vimukthi-git) [reply frame/composable-traits/src/dex.rs]: Unfortunately, there is no design document, but I also explained to you the same requirement before. We have ongoing discussions on supporting multi-asset stable swap soon. So this is not hypothetical. Since assetIds are expected to be provided through these functions, so for now, we can assert the order of assets provided as base first for now. Later we can switch the implementation easily.

Comment (vimukthi-git) [reply frame/composable-traits/src/dex.rs]: >Also I request to have small PRs which has changes related to the subject of given click-up ticket.

I understand, but this function signature is being changed in the current PR, so I see that it's not going to cause a change in PR scope.

**Review by KaiserKarel — CHANGES_REQUESTED** [review]
Would want to see the discussion regarding multi assets pool resolved before merging this.

File: frame/composable-traits/src/dex.rs
Comment (dzmitry-lahoda) [reply frame/composable-traits/src/dex.rs]: @vimukthi-git 
```
We need to introduce asset Ids (lists of assets) for these types of calls, names like base and quote
``` 
this is future requirements, came after current requirement to get things done. for me it feels by process should be separate PR

File: frame/composable-traits/src/dex.rs
Comment (dzmitry-lahoda) [inline frame/composable-traits/src/dex.rs:46]: docs should be canged too. so may be in other PR not to block merge. i am very agains adding new work into already decide scope until work doen is useles. it is not useless in this case.

File: frame/composable-traits/src/dex.rs
Comment (vimukthi-git) [reply frame/composable-traits/src/dex.rs]: @dzmitry-lahoda I do not consider this a new requirement nor was it not known before about this concept. You can refer this comment in a previous PR https://github.com/ComposableFi/composable/pull/1093#discussion_r886559668 where this same thing was explained.

Comment (vivekvpandya) [reply frame/composable-traits/src/dex.rs]: There is already a task for assessment https://app.clickup.com/t/2gh7cjb . I changed functions to use Vec<> but as we see there is also possibilities of using HashMap. IMHO these kind of discussion should be part of design. Also when we complete 3 pool support in pallet then FE also need to use it. So even for them it is a rework.

Comment (vimukthi-git) [reply frame/composable-traits/src/dex.rs]: @KaiserKarel solution makes sense. Only requirement from my end is to prevent us from assuming two assets at the interface level from now until we have a design/tasks to implement multi asset.

File: frame/composable-traits/src/dex.rs
Comment (vivekvpandya) [reply frame/composable-traits/src/dex.rs:46]: updated.

**Review by dzmitry-lahoda — DISMISSED** [review]

**Review by KaiserKarel — APPROVED** [review]

File: frame/pablo/rpc/src/lib.rs
Comment (benluelo) [inline frame/pablo/rpc/src/lib.rs:45]: these need to be camel case: pablo_simulateRemoveLiquidity (same for add rpc)

File: frame/pablo/src/lib.rs
Comment (benluelo) [inline frame/pablo/src/lib.rs:169]: perhaps move this to the bottom of the file? feels out of place here

File: runtime/dali/src/lib.rs
Comment (benluelo) [inline runtime/dali/src/lib.rs:1306]:
```suggestion
			let min_expected_amounts: BTreeMap<_, _> = min_expected_amounts.iter().map(|(k, v)| (k.0, v.0)).collect();
```

File: runtime/dali/src/lib.rs
Comment (benluelo) [inline runtime/dali/src/lib.rs:1289]:
```suggestion
			let amounts: BTreeMap<_, _> = amounts.iter().map(|(k, v)| (k.0,v.0)).collect();
```

File: frame/composable-traits/src/dex.rs
Comment (hussein-aitlahcen) [inline frame/composable-traits/src/dex.rs:56]: Why do we add another function for that, RPC are readonly so you can directly use the usual add/remove liquidity function from a RPC to dry run?

Comment (vivekvpandya) [reply frame/composable-traits/src/dex.rs:56]: I was not aware about that. So I added a similar function which just check if balances are enough on all sides.

Comment (vivekvpandya) [reply frame/composable-traits/src/dex.rs:56]: @hussein-aitlahcen so if I call a function which has #transactional from RPC , it will not change any thing in storage ?

Comment (vimukthi-git) [reply frame/composable-traits/src/dex.rs:56]: I also would like to know a definitive answer to this question :) , what happens when an RPC call is made that updates storage? Does it avoid any storage updates if there is no explicit initiation of an extrinsic?

Comment (hussein-aitlahcen) [reply frame/composable-traits/src/dex.rs:56]: It's a readonly operation that is not going to do anthing to the storage. So we are free to declare RPC that call functions called by extrinsics safely.

Comment (hussein-aitlahcen) [reply frame/composable-traits/src/dex.rs:56]: We need to be careful though as a node can be ddos, but if it's a RPC only node we can probably scale it. Computing the liquidity is quite fast

Comment (hussein-aitlahcen) [reply frame/composable-traits/src/dex.rs:56]: By the way, you should check out RPCs:
```
    system_dryRun
    system_dryRunAt
```

Comment (vivekvpandya) [reply frame/composable-traits/src/dex.rs:56]: @saadjhk could you please verify https://polkadot.js.org/docs/substrate/rpc/#dryrunextrinsic-bytes-at-blockhash-applyextrinsicresult if that can satisfy your needs?

**Review by KaiserKarel — APPROVED** [review]

**Review by PoisonPhang — DISMISSED** [review]

**Review by vimukthi-git — APPROVED** [review]
LGTM, though I see there is a question regarding proper impl of dry runs.

File: frame/composable-traits/src/dex.rs
Comment (saadjhk) [reply frame/composable-traits/src/dex.rs:56]: @hussein-aitlahcen those RPCs require signed extrinsics, FE will have to prompt users for signatures for each dryrun

**Review by KaiserKarel — DISMISSED** [review]

**Comment by vivekvpandya** [comment]
> 

@benluelo 

> This PR was included in a batch that successfully built, but then failed to merge into main. It will not be retried.
> 
> Additional information:
> 
> ```json
> {"message":"Waiting on code owner review from ComposableFi/parachain-rpc.","documentation_url":"https://docs.github.com/articles/about-protected-branches"}
> ```

**Comment by KaiserKarel** [comment]
@vivekvpandya You might need to reopen this PR, as Github kind of messes up required reviewers with team changes.

**Pull request closed without merging**
